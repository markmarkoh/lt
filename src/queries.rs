use graphql_client::GraphQLQuery;

type DateTime = String;

#[derive(Debug, Default, GraphQLQuery)]
#[graphql(
    schema_path = "src/schemas/linear.graphql",
    query_path = "src/queries/basic.graphql",
    response_derives = "serde::Serialize,Default,Debug,Clone"
)]
pub struct MyIssuesQuery;


