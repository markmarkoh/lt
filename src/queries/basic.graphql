fragment IssueFragment on Issue {
      title
      identifier
      state {
        name
        color
        type
      }
      url
      assignee {
        isMe
        displayName
      }
      creator {
        isMe
        displayName
      }
      estimate
      project {
        name
        icon
        color
      }
      createdAt
      priorityLabel
      priority
      labels {
        edges {
          node {
            color
            name
          }
        }
      }
      branchName
      description
}

query MyIssuesQuery {
    issues(
      filter: {
        assignee: { isMe: { eq: true } }
        state: { type: { in: ["started", "backlog", "unstarted"] } }
      },
      sort: {
        priority:  {
           order: Descending
        }
    }
  ) {
    nodes {
      ...IssueFragment
    }
  }
}
