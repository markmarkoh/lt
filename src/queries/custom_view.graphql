fragment IssueFragment on Issue {
      title
      identifier
      state {
        name
        color
        type
      }
      url
      assignee {
        isMe
        displayName
      }
      creator {
        isMe
        displayName
      }
      estimate
      project {
        name
        icon
        color
      }
      createdAt
      priorityLabel
      priority
      labels {
        edges {
          node {
            color
            name
          }
        }
      }
      branchName
      description
}

query CustomViewQuery($custom_view_id: String!) {
  customView(id: $custom_view_id) {
    issues {
      nodes {
        ...IssueFragment
      }
    }
  }
}
