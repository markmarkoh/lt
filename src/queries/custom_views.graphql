fragment ViewFragment on CustomView {
  name
  slugId
  color
  icon
  id
}
query CustomViewsQuery {
  customViews {
    nodes {
      ...ViewFragment
    }
  }
}
