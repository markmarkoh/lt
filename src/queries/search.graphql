fragment IssueFragment on IssueSearchResult {
      title
      identifier
      state {
        name
        color
        type
      }
      url
      assignee {
        isMe
        displayName
      }
      creator {
        isMe
        displayName
      }
      estimate
      project {
        name
        icon
        color
      }
      createdAt
      priorityLabel
      priority
      labels {
        edges {
          node {
            color
            name
          }
        }
      }
      branchName
      description
}

query SearchQuery($term: String!) {
  searchIssues(term: $term) {
    nodes {
      ...IssueFragment
    }
  }
}
