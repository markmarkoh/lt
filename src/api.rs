use graphql_client::{GraphQLQuery, Response};
use reqwest::Client;
use std::error::Error;

#[derive(Debug, Default)]
pub struct LinearClient {
    endpoint: String,
    client: Client,
}

impl LinearClient {
    pub fn new(api_key: String) -> Result<Self, Box<dyn Error>> {
        let endpoint = String::from("https://api.linear.app/graphql");
        let client = Client::builder()
            .user_agent("graphql-rust/0.10.0")
            .default_headers(
                std::iter::once((
                    reqwest::header::AUTHORIZATION,
                    reqwest::header::HeaderValue::from_str(&api_key.to_string()).unwrap(),
                ))
                .collect(),
            )
            .build()?;

        Ok(Self {
            client,
            endpoint,
        })
    }

    pub async fn query<T: GraphQLQuery>(&self, _query: T, variables: T::Variables) -> Result<T::ResponseData, Box<dyn Error>> {
        let var = T::build_query(variables);
        let res = self.client.post(&self.endpoint).json(&var).send().await?;
        let response_body: Response<T::ResponseData> = res.json().await?;
        Ok(response_body.data.unwrap())
    }
}

/*#[cfg(test)]
#[tokio::test]
mod tests {
    use super::*;
    use crate::queries::*;
    
    #[test]
    async fn hey() -> Option<()> {
        let linear_api_token =
            std::env::var("LINEAR_API_TOKEN").expect("Missing LINEAR_API_TOKEN env var");
        let client = LinearClient::new(linear_api_token).unwrap();
        let variables = example_query::Variables {};
        
        println!("Oka");
        let thi = client.query(ExampleQuery, variables).await?;
        println!("Oka2");
        if let Ok(v) = thi {
            println!("Okay!");
        }
        Some(())

        /*if let Ok(v) = client.query(ExampleQuery, variables) {
            assert_eq!(v.user_settings.id, "ad6eb6f5-4a44-440f-94aa-d3a9adcd0be9");
            for node in v.issues.nodes {
                assert!(!node.title.is_empty());
                println!("Call again: {:?}", node.title);
            }
        }*/
    }
}*/
