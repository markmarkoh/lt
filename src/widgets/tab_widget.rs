use crossterm::event::{Event, KeyCode, KeyEventKind};
use ratatui::buffer::Buffer;
use ratatui::{
    layout::Rect,
    widgets::{Tabs, Widget},
};
use std::sync::{Arc, RwLock};

use crate::TabChangeEvent;
use crate::api::LinearClient;
use crate::queries::{CustomViewsQuery, custom_views_query};

#[derive(Debug, Clone)]
pub struct TabWidget {
    state: Arc<RwLock<TabWidgetState>>,
}

#[derive(Debug, Clone)]
struct TabWidgetState {
    selected_index: usize,
    tabs: Vec<Tab>,
}

impl Default for TabWidget {
    fn default() -> Self {
        TabWidget {
            state: Arc::new(RwLock::new(TabWidgetState {
                selected_index: 0,
                tabs: vec![Tab {
                    title: String::from("My Issues"),
                    custom_view: None,
                }],
            })),
        }
    }
}

#[derive(Debug, Clone, Default)]
struct Tab {
    title: String,
    custom_view: Option<custom_views_query::ViewFragment>,
}

impl TabWidget {
    pub fn run(&self) {
        let this = self.clone();
        tokio::spawn(this.fetch());
    }

    async fn fetch(self) {
        let linear_api_token =
            std::env::var("LINEAR_API_TOKEN").expect("Missing LINEAR_API_TOKEN env var");
        let client = LinearClient::new(linear_api_token).unwrap();
        let variables = custom_views_query::Variables {};
        match client.query(CustomViewsQuery, variables).await {
            Ok(data) => {
                let mut state = self.state.write().unwrap();
                for custom_view in data.custom_views.nodes.iter() {
                    state.tabs.push(Tab {
                        title: custom_view.name.clone(),
                        custom_view: Some(custom_view.clone()),
                    });
                }
            }
            Err(e) => {
                panic!("Error {:#?}", e);
                //return;
            }
        }
        //self.set_loading_state(LoadingState::Loaded);
    }
    pub fn next(&self) {
        let mut state = self.state.write().unwrap();
        if state.selected_index < state.tabs.len() - 1 {
            state.selected_index += 1;
        }
    }

    pub fn prev(&self) {
        let mut state = self.state.write().unwrap();
        if state.selected_index > 0 {
            state.selected_index -= 1;
        }
    }

    pub fn handle_event(&self, event: &Event) -> crate::TabChangeEvent {
        if let Event::Key(key) = event {
            if key.kind == KeyEventKind::Press {
                match key.code {
                    KeyCode::Tab => {
                        self.next();
                        let state = self.state.read().unwrap();
                        match &state.tabs[state.selected_index].custom_view {
                            Some(custom_view) => {
                                return TabChangeEvent::FetchCustomViewIssues(custom_view.clone());
                            }
                            _ => return TabChangeEvent::FetchMyIssues,
                        }
                    }
                    KeyCode::BackTab => {
                        self.prev();
                        let state = self.state.read().unwrap();
                        match &state.tabs[state.selected_index].custom_view {
                            Some(custom_view) => {
                                return TabChangeEvent::FetchCustomViewIssues(custom_view.clone());
                            }
                            _ => return TabChangeEvent::FetchMyIssues,
                        }
                    }
                    _ => return TabChangeEvent::None,
                }
            }
        }
        TabChangeEvent::None
    }
}

impl Widget for &TabWidget {
    fn render(self, area: Rect, buf: &mut Buffer) {
        Tabs::new(
            self.state
                .read()
                .unwrap()
                .tabs
                .iter()
                .map(|tab| tab.title.clone())
                .collect::<Vec<String>>(),
        )
        .select(self.state.read().unwrap().selected_index)
        .padding(" ", " ")
        .divider("  ")
        .render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use std::sync::{Arc, RwLock};

    use crossterm::event::{KeyCode, KeyEventKind, KeyEventState, KeyModifiers};
    use insta::assert_snapshot;
    use ratatui::{Terminal, backend::TestBackend};

    use crate::{
        TabChangeEvent,
        queries::custom_views_query,
        widgets::TabWidget,
    };

    use super::{Tab, TabWidgetState};

    fn create_key_event(code: KeyCode) -> crossterm::event::Event {
        crossterm::event::Event::Key(crossterm::event::KeyEvent {
            code,
            kind: KeyEventKind::Press,
            modifiers: KeyModifiers::empty(),
            state: KeyEventState::empty(),
        })
    }

    #[test]
    fn test_empty_state() {
        let app = TabWidget::default();
        let mut terminal = Terminal::new(TestBackend::new(50, 2)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());

        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(ev, TabChangeEvent::FetchMyIssues);

        let ev = app.handle_event(&create_key_event(KeyCode::Char('j')));
        assert_eq!(ev, TabChangeEvent::None);
    }

    #[test]
    fn test_multi_tabs() {
        let app = TabWidget {
            state: Arc::new(RwLock::new(TabWidgetState {
                selected_index: 0,
                tabs: vec![
                    Tab {
                        title: String::from("My Issues"),
                        custom_view: None,
                    },
                    Tab {
                        title: String::from("Custom A"),
                        custom_view: Some(custom_views_query::ViewFragment {
                            slug_id: Some("sluga".into()),
                            id: "sluga".into(),
                            name: "Custom A".into(),
                        }),
                    },
                    Tab {
                        title: String::from("Custom B"),
                        custom_view: Some(custom_views_query::ViewFragment {
                            slug_id: Some("slugb".into()),
                            id: "slugb".into(),
                            name: "Custom B".into(),
                        }),
                    },
                ],
            })),
        };

        let mut terminal = Terminal::new(TestBackend::new(50, 2)).unwrap();
        terminal
            .draw(|frame| frame.render_widget(&app, frame.area()))
            .unwrap();

        assert_snapshot!(terminal.backend());

        // should go to  tab a
        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(ev, TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
            name: "sluga".into(),
            slug_id: Some("sluga".into()),
            id: String::from("sluga")
        }));

        // then to b
        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(ev, TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
            name: "slugb".into(),
            slug_id: Some("slugb".into()),
            id: String::from("slugb")
        }));

        // but not passed b
        let ev = app.handle_event(&create_key_event(KeyCode::Tab));
        assert_eq!(ev, TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
            name: "slugb".into(),
            slug_id: Some("slugb".into()),
            id: String::from("slugb")
        }));

        // then back to a 
        let ev = app.handle_event(&create_key_event(KeyCode::BackTab));
        assert_eq!(ev, TabChangeEvent::FetchCustomViewIssues(custom_views_query::ViewFragment {
            name: "sluga".into(),
            slug_id: Some("sluga".into()),
            id: String::from("sluga")
        }));

        let ev = app.handle_event(&create_key_event(KeyCode::BackTab));
        assert_eq!(ev, TabChangeEvent::FetchMyIssues);


    }
}
