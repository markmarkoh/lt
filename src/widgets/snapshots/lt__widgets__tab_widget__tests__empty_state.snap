---
source: src/widgets/tab_widget.rs
expression: terminal.backend()
---
" My Issues                                        "
"                                                  "
